Source/DafnyCore/Rewriters/AllOpaqueRevealStmtInserter.cs

```diff
@@ -0,0 +1,531 @@
+                  var newExpr = new StmtExpr(expr.Witness.Tok, revealStmt, origExpr) {
+                    Type = origExpr.Type
+                  };
+                  expr.Witness = newExpr;
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private class GraphTraversalVertex {
+    public readonly Graph<ICallable>.Vertex Vertex;
+    public readonly bool Local;
+
+    public List<NameSegment> NamePath;
+    public List<ModuleDefinition> ModulePath;
+
+    public GraphTraversalVertex(Graph<ICallable>.Vertex vertex, bool local, List<NameSegment> namePath, List<ModuleDefinition> modulePath) {
+      Vertex = vertex;
+      Local = local;
+      NamePath = namePath;
+      ModulePath = modulePath;
+    }
+
+    public override bool Equals(object obj) {
+      var other = obj as GraphTraversalVertex;
+      return other?.Vertex.N.FullSanitizedName == Vertex.N.FullSanitizedName;
+    }
+
+    public override int GetHashCode() {
+      return HashCode.Combine(Vertex.N.FullSanitizedName);
+    }
+  }
+
+  private void AddMethodReveals(Method m, ErrorReporter reporter) {
+    Contract.Requires(m != null);
+
+    bool autoRevealDeps = Attributes.ContainsBoolAtAnyLevel(m, "autoRevealDependencies", true);
+
+    var currentClass = m.EnclosingClass;
+    List<RevealStmt> addedReveals = new List<RevealStmt>();
+
+    using IEnumerator<FunctionReveal> enumerator = GetEnumerator(m, currentClass, m.SubExpressions);
+    while (enumerator.MoveNext()) {
+      var func = enumerator.Current;
+      var revealStmt = BuildRevealStmt(func, m.Tok, m.EnclosingClass.EnclosingModuleDefinition);
+
+      if (revealStmt is not null) {
+        addedReveals.Add(revealStmt);
+        if (autoRevealDeps) {
+          if (m is Constructor c) {
+            c.BodyInit.Insert(0, revealStmt);
+          } else {
+            m.Body.Body.Insert(0, revealStmt);
+          }
+        }
+      }
+    }
+
+    if (autoRevealDeps) {
+      if (m.Req.Any() || m.Ens.Any()) {
+        Expression reqExpr = new LiteralExpr(m.Tok, true);
+        reqExpr.Type = Type.Bool;
+        foreach (var revealStmt0 in addedReveals) {
+          reqExpr = new StmtExpr(reqExpr.tok, revealStmt0, reqExpr);
+          reqExpr.Type = Type.Bool;
+        }
+
+        m.Req.Insert(0, new AttributedExpression(reqExpr));
+      }
+    }
+
+    if (addedReveals.Any()) {
+      reporter.Message(MessageSource.Rewriter, ErrorLevel.Info, null, m.tok, addedReveals.Count + " reveal stmt(s) inserted: \n" + RenderRevealStmts(addedReveals));
+    }
+  }
+
+  private void AddFunctionReveals(Function f, ErrorReporter reporter) {
+    Contract.Requires(f != null);
+
+    bool autoRevealDeps = Attributes.ContainsBoolAtAnyLevel(f, "autoRevealDependencies", true);
+
+    var currentClass = f.EnclosingClass;
+    List<RevealStmt> addedReveals = new List<RevealStmt>();
+
+    using IEnumerator<FunctionReveal> enumerator = GetEnumerator(f, currentClass, f.SubExpressions);
```

Comment:
I see you using this using/while loop idiom a lot. Could you perhaps move this iteration logic somewhere else and just call a higher-order function? That might make things more legible and maintainable.